File: np-front/src/search/test-n-puzzle.ts
import { NTiles, isEven } from "./util";

// https://www.geeksforgeeks.org/check-instance-8-puzzle-solvable/
// https://www.geeksforgeeks.org/check-instance-15-puzzle-solvable/

export class TestNPuzzle {

  private nSize = 0;

  constructor(private readonly startPuzzle: NTiles, private readonly goalPuzzle: NTiles) {
    this.nSize = goalPuzzle.length;
  }
  private _getInversionsCount(arr: NTiles) {
    let inv_count = 0;
    for (let row = 0; row < this.nSize; row++) {
      for (let col = 0; col < this.nSize; col++) {
        if (arr[row][col] === 0) {
          continue;
        }

        // جستجوی کاشی های جلویی
        let j = col + 1;
        for (let i = row; i < this.nSize; i++) {
          for (; j < this.nSize; j++) {
            if (arr[i][j] && arr[i][j] < arr[row][col])
              inv_count++;
          }
          j = 0;
        }
      }
    }
    return inv_count;
  }

  private _findBlankPosition(puzzle: NTiles) {
    for (let i = 0; i < this.nSize; i++)
      for (let j = 0; j < this.nSize; j++)
        if (puzzle[i][j] === 0)
          return i + 1;
    return 0;
  }

  private _calcInversions(puzzle: NTiles) {
    let invCount = this._getInversionsCount(puzzle);
    if (isEven(this.nSize)) {
      invCount += this._findBlankPosition(puzzle);
      return invCount;
    }
    return invCount;
  }

  calcSolvable() {
    const start = this._calcInversions(this.startPuzzle);
    const goal = this._calcInversions(this.goalPuzzle);
    const result = {
      start,
      goal,
      startRem: start % 2,
      goalRem: goal % 2,
      canSolve: false
    };
    result.canSolve = (result.startRem === result.goalRem);
    return result;
  }
}

File: cs-console/TestNPuzzle.cs
public class TestNPuzzle
{
    private int nSize = 0;
    private readonly short[][] startPuzzle;
    private readonly short[][] goalPuzzle;

    public TestNPuzzle(short[][] startPuzzle, short[][] goalPuzzle)
    {
        this.nSize = goalPuzzle.Length;
        this.startPuzzle = startPuzzle;
        this.goalPuzzle = goalPuzzle;
    }
    private int _getInversionsCount(short[][] arr)
    {
        var inv_count = 0;
        for (var row = 0; row < this.nSize; row++)
        {
            for (var col = 0; col < this.nSize; col++)
            {
                if (arr[row][col] == 0)
                {
                    continue;
                }

                // جستجوی کاشی های جلویی
                var j = col + 1;
                for (var i = row; i < this.nSize; i++)
                {
                    for (; j < this.nSize; j++)
                    {
                        if (arr[i][j] != 0 && arr[i][j] < arr[row][col])
                            inv_count++;
                    }
                    j = 0;
                }
            }
        }
        return inv_count;
    }

    private short _findBlankPosition(short[][] puzzle)
    {
        for (short i = 0; i < this.nSize; i++)
            for (short j = 0; j < this.nSize; j++)
                if (puzzle[i][j] == 0)
                    return (short)(i + 1);
        return 0;
    }

    private int _calcInversions(short[][] puzzle)
    {
        var invCount = this._getInversionsCount(puzzle);
        if (this.nSize % 2 == 0)
        {
            invCount += this._findBlankPosition(puzzle);
            return invCount;
        }
        return invCount;
    }

    public TestResult calcSolvable()
    {
        var start = this._calcInversions(this.startPuzzle);
        var goal = this._calcInversions(this.goalPuzzle);
        TestResult result = new TestResult(start, goal);
        return result;
    }
}
